Common BLE API

jimmo opened the pull request:
This PR is based on #4589 by @aykevl and #4893 by @dpgeorge 

It implements a Python API for working with BLE devices in both central and peripheral role (including scanning and advertising), and is implemented with Nimble for STM32. I have WIP ESP32 support (based on #4859, but including central role) almost ready to go, although it might be an option to use Nimble on ESP32 instead. NRF is also in progress using the SD (but could potentially use Nimble also!).

The API is lower-level than #4589 -- it more closely matches the underlying BLE spec. The idea is that specific use cases (beacon, beacon detector, UART, etc) can be implemented with higher-level Python wrappers.

Some outstanding questions:
 - Memory management for Nimble. At the moment a soft reset will crash because Nimble allocates on the uPy heap.
 - Naming. Unsure about the module (`bluetooth`), class (`Bluetooth`), and method names (e.g. 'characteristics' vs 'chrs'). I hope that a future BT Classic implementation can share some of this so some thought required about whether to rename anything to be BLE specific.
 - Address types (private/public/static/random) and configuration.

Registering a service (e.g. NUS):
```
import bluetooth
bt = bluetooth.Bluetooth()
bt.active(True)

UART_SERVICE = bluetooth.UUID('6E400001-B5A3-F393-E0A9-E50E24DCCA9E')
UART_TX = (bluetooth.UUID('6E400003-B5A3-F393-E0A9-E50E24DCCA9E'), bluetooth.FLAG_READ|bluetooth.FLAG_NOTIFY,)
UART_RX = (bluetooth.UUID('6E400002-B5A3-F393-E0A9-E50E24DCCA9E'), bluetooth.FLAG_WRITE,)

tx, rx = bt.gatts_add_service(UART_SERVICE, (UART_TX, UART_RX,))
```

Scanning:
```
def bt_irq(event, data):
    if event == IRQ_SCAN_RESULT:
        addr_type, addr, connectable, rssi, adv_data = data
        print('Scan result', adv_decode_name(adv_data))
    elif event == IRQ_SCAN_COMPLETE:
        print('Scan complete')

bt.irq(bt_irq, IRQ_ALL)
bt.scan(2000)  # Scan for 2000 ms
```

Advertising:
```
# The encoding will be available as Python helpers.
bt.advertise(100, adv_encode_ble_only() + adv_encode_name('pybd-jimmo'))
```

The methods available on the `Bluetooth` object are:

```
# General
active([enable])
config(name) # e.g. config('mac')
irq(handler, trigger)

# Peripheral
gatts_add_service(uuid, ( (chr_uuid, flags), ... ) )
advertise(interval_ms, data)
gatts_notify(conn_handle, value_handle [, data])  # Optional data
gatts_read(conn_handle, value_handle)
gatts_write(conn_handle, value_handle, data)

# Central
scan(timeout_ms)  # None to stop scanning.
connect(addr_type, addr)
disconnect(conn_handle)
gattc_discover_services(conn_handle)
gattc_discover_characteristics(conn_handle, start_handle, end_handle)
gattc_discover_descriptors(conn_handle, start_handle, end_handle)
gattc_read(conn_handle, value_handle)
gattc_write(conn_handle, value_handle, data)
```

All events are raised via the irq handler. On all ports this is a "soft" handler (i.e. you can allocate memory). The events (and their corresponding data tuple) are:

```
# Peripheral role
IRQ_CENTRAL_CONNECT  (conn_handle, addr_type, addr,)
IRQ_CENTRAL_DISCONNECT  (conn_handle,)
IRQ_CHARACTERISTIC_WRITE  (conn_handle, attr_handle,)

# Central role
IRQ_SCAN_RESULT  (addr_type, addr, connectable, rssi, adv_data,)
IRQ_SCAN_COMPLETE  
IRQ_PERIPHERAL_CONNECT  (conn_handle, addr_type, addr,)
IRQ_PERIPHERAL_DISCONNECT  (conn_handle,)
IRQ_PERIPHERAL_SERVICE_RESULT  (conn_handle, start_handle, end_handle, uuid,)
IRQ_PERIPHERAL_CHARACTERISTIC_RESULT  (conn_handle, def_handle, value_handle, properties, uuid,)
IRQ_PERIPHERAL_DESCRIPTOR_RESULT  (conn_handle, descriptor_handle, uuid,)
IRQ_PERIPHERAL_READ_RESULT  (conn_handle, value_handle, char_data,)
IRQ_PERIPHERAL_WRITE_STATUS  (conn_handle, value_handle, status,)
IRQ_PERIPHERAL_NOTIFY  (conn_handle, value_handle, notify_data,)
IRQ_PERIPHERAL_INDICATE  (conn_handle, value_handle, notify_data,)
```

As a peripheral (i.e. GATT server), you can use the `gatts_*` methods to read/write/notify/indicate a characteristic (using the characteristic handle returned from `gatts_add_service`, and optionally a conn_handle from IRQ_CENTRAL_CONNECT)

As a central, you can obtain a connection handle and characteristic handle using the GATT client methods (`gattc_*`) and use gattc_read/write, or receive notifications by IRQ_PERIPHERAL_NOTIFY.

dpgeorge committed:
drivers/cyw43: Add low-level CYW43xx Bluetooth HCI UART driver.

dpgeorge committed:
lib/mynewt-nimble: Add Apache mynewt nimble as a submodule.

Tag nimble_1_1_0_tag.

dpgeorge committed:
extmod/nimble: Add nimble bindings.

jimmo committed:
extmod/modbluetooth: Add low-level Python BLE API.

jimmo committed:
stm32/boards/PYBD: Enable BLE for Pyboard D.

jimmo committed:
py/ringbuf: Add helpers for put16/get16.

andrewleech commented:
If you want a reference to look at re by classic api compatibility, zephyr supports some classic in their stack, so perhaps their application api would be useful: 
https://docs.zephyrproject.org/1.12.0/subsystems/bluetooth/bluetooth.html
https://docs.zephyrproject.org/latest/guides/bluetooth/overview.html

https://docs.zephyrproject.org/latest/reference/bluetooth/index.html

While that doesn't list a2dp there is code for it in the repo: https://docs.zephyrproject.org/apidoc/latest/a2dp_8h.html

As first skim though, it looks like there's a separate C interface/functions for each classic profile, which sits alongside the ble gattc etc top level functions. There may not be much compatibility/interacting between them at all.

dpgeorge commented:
Thanks @jimmo for the hard work.  I'm fully supportive of the approach to the API that is proposed here, building and improving on the work done by @aykevl .

The API is flat (a set of functions, no classes), minimal and powerful, reflecting closely the actual BLE spec.  This means it can be implemented with minimal code, yet allows to do (almost) anything that the BLE spec allows.

Using the API here for the core functionality, it's then possible to write wrappers in Python to expose a more user-friendly API, eg to provide a simple way to make peripherals like heart-rate monitors.  It's also possible to make wrappers to provide compatibility with other BLE APIs, eg to match how BLE would work on unix (compatibility with `bluepy` or `pygatt`), or on CircuitPython.

andrewleech commented:
Hi @jimmo, Have you already written some example helper functions you mentioned for advertising data?
`adv_encode_ble_only() + adv_encode_name('pybd-jimmo')`

andrewleech committed:
extmod/modbluetooth_nimble: Use random addr if public isn't available.

andrewleech commented:
@jimmo The changes I made to also support generic HCI radio (nrf52 running zephyr) are on my branch here, on top of your changes: https://github.com/andrewleech/micropython/tree/common-ble-api-anl

andrewleech commented on extmod/modbluetooth_nimble.c:
```diff
@@ -0,0 +1,690 @@
+            entry = MP_OBJ_TO_PTR(elem->value);
+            os_mbuf_append(ctxt->om, entry->data, entry->data_len);
+
+            return 0;
```
It's common on other bluetooth stacks to allow for a callback here when a read request comes in. On NRF devices this is exposed as a callback event BLE_GATTS_EVT_RW_AUTHORIZE_REQUEST.

It's officially used to provide the application with a chance to authorise or deny access to the characteristic based on whether it's paired etc.

It's also commonly used to allow the application to change the value of the characteristic each read. This can be then used to stream data out the characteristic reliably, instead of notifications which are far more unreliable as they don't have the same handshaking to ensure they get through.

The problem in this case is your other events are soft handlers, whereas this would need to be hard / inline to allow the callback to change the return value of this function.

pi-anl committed:
extmod/modbluetooth: Allow MP_BLUETOOTH_MAX_ATTR_SIZE in board config.

andrewleech commented:
It appears to only support one service currently?

If I run `bt.gatts_add_service()` multiple times for each service, I get printouts for for each svc & chr
```
gatts_register_cb: svc uuid=d0710dac handle=1
gatts_register_cb: chr uuid=d0710d58 def_handle=2 val_handle=3
gatts_register_cb: svc uuid=d071155c handle=5
gatts_register_cb: chr uuid=d0711448 def_handle=6 val_handle=7
gatts_register_cb: chr uuid=d0711460 def_handle=8 val_handle=9
gatts_register_cb: chr uuid=d0711478 def_handle=10 val_handle=11
gatts_register_cb: chr uuid=d0711490 def_handle=12 val_handle=13
gatts_register_cb: chr uuid=d07114a8 def_handle=14 val_handle=15
gatts_register_cb: chr uuid=d07114c0 def_handle=16 val_handle=17
```
But only the last service registered shows on mv nrf ble scanner app once connected. 
If I reverse the order they're configured, I get the other one shown.

andrewleech commented:
FYI I've made a quick python util function to create appropriate advertising data to include `device name`. I'll eventually extend it to handle services and scan response as well, but for now I've chucked a snapshot here: https://gist.github.com/andrewleech/49c2a54313f57e544dc1b3a3360d6dc3

andrewleech commented:
I've added support for multiple services to the head of my extension branch, see in https://github.com/andrewleech/micropython/commit/72b2d179d2560a1c09716758a3cb2b3cb3459768
I do have quite a few other smaller changes on that branch too: https://github.com/andrewleech/micropython/commits/common-ble-api-anl

It required a fair bit of turning the code around, not sure it's the best option. 
Now, Bluetooth().active(True) must only be called after all your services have been registered, this is what triggers the stack startup.
Bluetooth().handles() can then be called to get a dict of all the service uuid's keyed to a tuple of char uuids with a buffer of the matching handles.

It's quite possibly better to make the one add service function take a larger structure of all the services and their matching chars in the one call, rather than having the extra handles tracking dict I've got in my attempt.

rf-ir-learner commented:
Hello, I am new to BLE on ESP32 and I'm wondering if I could get some clarification on this. I am quite excited by the prospect of being able to use BLE with Micropython, since so many of the sensors these days seem to have really long battery life with BLE (like a year or more on a button battery).

It seems like we can use some basic BLE but I'm a bit confused as to how it is implemented. I need to set up an ESP32 board as a client to monitor a few BLE sensors, nothing terribly complicated.

So is BLE contained within a version of Micropython that I can just burn and install on my board? If so, is there a particular version (link?) I need to download?

Or is the BLE component a separate library module that I would install in the LIB sub-folder on my ESP32? If so, is there a link for that?

Or is there some other thing I would have to do (compile something)?

Any help around this would be greatly appreciated; these are probably really stupid questions, but I really am confused.

Regards, AB

andrewleech commented:
Hi @rf-ir-learner, this PR doesn't yet have support for esp32, the stm32 support has been proofed out first. 

I know Esp32 and others are being worked on, but for a feature as important as Bluetooth @jimmo and others are putting in a lot of effort behind the scenes planning it all to try to ensure its done right the first time.

dpgeorge commented:
> It's quite possibly better to make the one add service function take a larger structure of all the services and their matching chars in the one call

It's probably not wise to rely on the underlying BLE stack (nimble, nordic, bluedroid) being able to add services "later on".  In other words, the common denominator here would be to assume that all services must be registered at the same time.

In that case `gatts_add_service()` should be renamed to `gatts_register_services()` (or `gatss_set_services`) and take a list of what `gatts_add_services` previously accepted.  You could call `gatts_register_services` more than once (and after `bt.active(1)`) but each call would completely reset the available service list.  There could even be an extension in the future which didn't reset the list, eg `gatts_register_services(..., reset=False)`.

Alternatively, it might be possible to keep `gatts_add_service()` and automatically detect when all services were added, eg:
```
gatts_add_service(...) # first call to this func, so reset gatts before adding
gatts_add_service(...) # accumulate service
advertise(...) # service list must be done so register them, then advertise
```
But that seems a bit flaky.

dhalbert commented:
> the common denominator here would be to assume that all services must be registered at the same time.

I recently went through some API churn on this with `bleio` (now `_bleio`) for CircuitPython. My original API had all services passed at once to a peripheral, as @dpgeorge suggests. The tree of services, their characteristics, and the characteristic's descriptors was built up all at once and then passed in. As background, note that we are supporting peripheral and central mode, peripherals as clients, not just servers, and pairing/bonding.

One problem I encountered with this approach was that `Service`, `Characteristic`, and `Descriptor` had conventional constructors. So there would loose, unconnected objects during service tree construction. A user of the API might inadvertently reuse one of those attribute objects when constructing multiple services. To prevent that I ended up dropping the all-at-once addition and creating factory methods like `Service.add_to_peripheral()`, so that there were no unattached objects. I'm not completely happy with this either, but it's safer, and it removed the need for some bad-state checking internally. Here's the PR for that, with discussion: https://github.com/adafruit/circuitpython/pull/2092, and [here is a snapshot](https://github.com/adafruit/Adafruit_CircuitPython_BLE/blob/33d75f4ca8e52e964d060db1f046b16255b8428b/adafruit_ble/hid_server.py
) of a complicated use of the API, to build up a BLE HID service.

@tannewt has suggested that the service/characteristic/descriptor tree be described in a more declarative way, usable for both local and remote attributes. These service trees would then be passed to a peripheral or central object. In the case of a peripheral, the tree components would be instantiated and registered with the BLE stack. In the case of a central, the user would, for example, use a property in the declarative tree to set or get the value of a remote characteristic, and the internal logic would check that a matching characteristic had been found during discovery.

Because a BLE stack is so complicated, we decided to name the low-level module `_bleio` to indicate it should not be used by ordinary users, and so we can change its API between major versions of CircuitPython. We have a library layer (https://github.com/adafruit/Adafruit_CircuitPython_BLE) that will hide the details and provide common services and functionality.

(I could critique the BLE protocol, which has a multitude of role-changing chameleon objects: it is more of an n-dimensional basket-weave of orthogonal concepts than a "stack". But it is what it is and we have to figure out how to present it in a usable way.)

jimmo commented:
Thank you for the feedback @andrewleech & @dpgeorge.

I have updated this branch, incorporating all of the improvements from @andrewleech (with some modifications). I have also re-implemented adding multiple services based on @dpgeorge's suggestion. An example of adding multiple services now looks like:

```
bt = bluetooth.Bluetooth()
bt.active(True)

UART_SERVICE = bluetooth.UUID('6E400001-B5A3-F393-E0A9-E50E24DCCA9E')
UART_TX = (bluetooth.UUID('6E400003-B5A3-F393-E0A9-E50E24DCCA9E'), bluetooth.FLAG_READ|bluetooth.FLAG_NOTIFY,)
UART_RX = (bluetooth.UUID('6E400002-B5A3-F393-E0A9-E50E24DCCA9E'), bluetooth.FLAG_WRITE,)
HR_SERVICE = bluetooth.UUID(0x180D)
HR_CHAR = (bluetooth.UUID(0x2A37), bluetooth.FLAG_READ|bluetooth.FLAG_NOTIFY,)

((hr,), (uart_tx, uart_rx,),) = bt.gatts_register_services(
                                          (
                                           (HR_SERVICE, (HR_CHAR,),),
                                           (UART_SERVICE, (UART_TX, UART_RX,),),
                                          )
                                         )
```

This allows you to modify the service list multiple times.

dhalbert commented:
@jimmo I  see our comments just crossed, and that you create the tree in the call. If you add descriptors at some point, then you'll need further nesting in the tuple of tuples.

dhalbert commented:
For something like BLE HID, I have also had to add specification of security modes, and values for an attributes for an attribute's max length and whether it is fixed length. I also added the ability to specify an initial value for an attribute, which is common for read-only descriptor values.

jimmo commented:
> @jimmo I see our comments just crossed, and that you create the tree in the call. If you add descriptors at some point, then you'll need further nesting in the tuple of tuples.

Thanks for the feedback also @dhalbert -- sorry I just missed your first comment. Yes, descriptors would be an optional third entry in the characteristic tuple (and in the returned value handles tree), but this does start to get a bit messy. Not to mention other properties (default values, security, etc). (Edit: just saw your next comment :) Some thought required here - I will follow up with @dpgeorge )

We have some additional constraints due to using Nimble as the BLE stack, which requires all the services to be defined at once. See https://github.com/apache/mynewt-nimble/issues/556  I need to spend some more time looking into your implementation, but I'm not sure that it would be possible currently to provide the _bleio API on top of Nimble.

Obviously we'd prefer not to be forced into a particular API design by the underlying stack...

> Here's the PR for that, with discussion: adafruit#2092, and here is a snapshot of a complicated use of the API, to build up a BLE HID service.

Thank you -- this is really useful!

> Because a BLE stack is so complicated, we decided to name the low-level module _bleio to indicate it should not be used by ordinary users, and so we can change its API between major versions of CircuitPython. We have a library layer (https://github.com/adafruit/Adafruit_CircuitPython_BLE) that will hide the details and provide common services and functionality.

This was our intention too (this is our low-level API). Our goal initially was to avoid any classes at all (hence the tuples) and then match the BLE gatts/gattc operation directly. I hope it should be still possible to make this match at the high-level API.

> (I could critique the BLE protocol, which has a multitude of role-changing chameleon objects: it is more of an n-dimensional basket-weave of orthogonal concepts than a "stack". But it is what it is and we have to figure out how to present it in a usable way.)

:) Helps me feel better about how long this is taking!

dhalbert commented:
@jimmo Thanks for your reply and for your work on this API. Your comments and approach have also given me some new ideas.

>Our goal initially was to avoid any classes at all (hence the tuples) and then match the BLE gatts/gattc operation directly.

It's not so far from positional tuples to namedtuples to declarative classes: I think you may end up moving in that direction for convenience and readability as you need to add yet more properties for the attributes.

>We have some additional constraints due to using Nimble as the BLE stack, which requires all the services to be defined at once.

I only found out about the existence of Nimble midway through working on `bleio`. That's an important potential constraint.

I think a key thing may be to separate the declarative and operational attribute classes. I have been assuming they would be the same, but it causes more state to need to be kept internally to track the object life cycle, and whether it is local or remote. If you look at something like the iOS BLE API, there's almost a Cartesian product of classes for objects and their roles, and maybe we could avoid that.

jimmo commented:
I've added another commit to address Nimble memory management (root pointers are tracked for nimble malloc, and it works across soft reset).

jimmo commented:
I've implemented descriptor support and basic support for (gatts) writing of longer values. Fixed a few other crashes and cleaned up some debugging/logging code.

Now able to make the PYBD act as a HID keyboard for my Android phone.

andrewleech commented:
The other thing I haven't looked at yet is pairing & bonding. 

"Just works" Pairing will likely work already, though I'm not sure about nimble's "LE Secure" support for using it safely. For other pairing (eg. Pin) another immediate/interrupt style event will likely be needed.

Bonding support will presumably need since functionality exposed/wired up to write the bonding info to flash.

If anyone else if a little lost in the wild range of different ble security options, this blog is a good primer: https://medium.com/rtone-iot-security/deep-dive-into-bluetooth-le-security-d2301d640bfc

dpgeorge commented on extmod/modbluetooth.c:
```diff
@@ -0,0 +1,1186 @@
+#if MICROPY_PY_BLUETOOTH_ENABLE_CENTRAL_MODE
+    { MP_ROM_QSTR(MP_QSTR_connect), MP_ROM_PTR(&bluetooth_connect_obj) },
+    { MP_ROM_QSTR(MP_QSTR_scan), MP_ROM_PTR(&bluetooth_scan_obj) },
+#endif
```
Please indent #'s to match the level that they protect.

dpgeorge commented on extmod/modbluetooth.c:
```diff
@@ -0,0 +1,1186 @@
+#if !MICROPY_PY_BLUETOOTH_CALLBACK_ALLOC
+STATIC mp_obj_t bluetooth_invoke_irq(mp_obj_t none_in) {
+    // This is always executing in schedule context.
+    while (1) {
```
Cody style is to use `for (;;) {`

dpgeorge commented on extmod/modbluetooth.h:
```diff
@@ -0,0 +1,251 @@
+// Event codes for the IRQ handler.
+// Can also be combined to pass to the trigger param to select which events you are interested in.
+// Note this is currently stored in a uint16_t (trigger, event), so any further events will require widening these.
+#define MP_BLUETOOTH_IRQ_CENTRAL_CONNECT              (1 << 1)
```
What about `(1 << 0)`?

jimmo commented:
OK thanks for the review feedback - I've addressed all those points exactly as suggested and updated the branch.

Note that the event numbers are now shifted right by one so they correctly start at 1 (oops), so Python code will need to be updated.

jimmo commented:
I've pushed some more updates:

- Renamed the `Bluetooth` class to `BLE`, as discussed (offline). So it's now:

```python
import bluetooth
ble = bluetooth.BLE()
ble.active(True)
```

This makes it obvious to the reader that this is BLE, not classic, and classic functionality can later be added either to role-specific classes (e.g. A2DP) or perhaps a `Classic` class (or both).

- Clarified and standardised byte ordering for addresses. In the Python API they are now always big-endian. This matches the way they're presented "in the real world" (i.e. on device stickers), and in high-level UIs (e.g. NRF connect). @andrewleech I would be interested in feedback you have here.

- Minor fixes (by renaming the BLE class I went through and made a bunch of internal naming in modbluetooth.c more consistent), and also fixed the RSSI (it's now correctly in dBm). I've also completely removed the "allocating" version of the event handlers, it now always unconditionally uses the ringbuf. We may bring this back in the future for other ports. (But it seems likely that the NRF SD won't need it). I've also refactored the ringbuf extraction (saving some code space).

dpgeorge commented:
Thanks, I have had another look through this and it's pretty much ready to go in.

I already cherry picked to master @andrewleech 's stm32 commit to add optional RTS/CTS support on UART4, because it's independent to this PR.

When merging I would do the following reorganisation of files (let me know if you think this is not a good idea):
- move `ports/stm32/cywbt.?` to `drivers/cyw43/` (this low-level BT driver can be used for other ports, not just stm32)
- move `ports/stm32/nimble/*` to `extmod/nimble/` (as above, these nimble low-level bindings can be used for other ports, not just stm32)

Also there are some very minor code styling things to fix, which I'll do during merge.

dpgeorge commented on extmod/modbluetooth_nimble.c:
```diff
@@ -0,0 +1,890 @@
+        case BLE_GAP_EVENT_DISCONNECT:
+            // Disconnect.
+            // TODO: Include address? event->disconnect.conn.peer_id_addr.val
+            mp_bluetooth_gap_on_peripheral_disconnected(event->disconnect.conn.conn_handle);
```
Include address?

jonsmirl commented:
I've been giving this code a try on the esp32-idf 4. It seems to be working, but it's not clear to me exactly how to use it. Can you help out with example code to get the node to a point where a phone can see it on the Bluetooth screen?

jimmo commented on extmod/modbluetooth.c:
```diff
@@ -0,0 +1,1002 @@
+    { MP_ROM_QSTR(MP_QSTR_config), MP_ROM_PTR(&bluetooth_ble_config_obj) },
+    { MP_ROM_QSTR(MP_QSTR_irq), MP_ROM_PTR(&bluetooth_ble_irq_obj) },
+    // GAP
+    // TODO: Potentially rename these to gap_*.
```
I would prefer `gap_connect` etc, just to be consistent with the `gatts_` / `gattc_` of the other methods. What do you think?

dpgeorge replied:
gattc/gatts are necessary because of the C vs S; gap is not needed from this point of view, the names like "connect" are unique.  But, considering the low-level, BLE-spec-following nature of the API, it does make sense to use `gap_connect`, `gap_advertise` etc.  It's just a lot of characters in all the corresponding qstrs...

jimmo replied:
I've switched it to `gap_`. 16 extra bytes of qstrs. But... 400 bytes saved by collapsing all the connect/disconnect event paths down to one, now that disconnect also includes the address.

jimmo commented:
@dpgeorge Thanks, I will address those comments (and move the files as suggested) over the weekend.

> I've been giving this code a try on the esp32-idf 4. It seems to be working, but it's not clear to me exactly how to use it. Can you help out with example code to get the node to a point where a phone can see it on the Bluetooth screen?

@jonsmirl I hope you found https://github.com/jimmo/micropython/tree/common-ble-api-esp32-nimble which is my WIP branch for exactly this. (This is why I recently added support for the IDF 4.x).

Here's a simple example of a heart rate monitor (tested on PYBD (this PR) and ESP32 (the linked branch)).

```python
import bluetooth
bt = bluetooth.BLE()   # This might still be bluetooth.Bluetooth in the ESP32 branch
bt.active(True)

HR_SERVICE = bluetooth.UUID(0x180D)
HR_CHAR = (bluetooth.UUID(0x2A37), bluetooth.FLAG_READ|bluetooth.FLAG_NOTIFY,)

((hr,),
) = bt.gatts_register_services(
                               (
                                (HR_SERVICE, (HR_CHAR,),),
                               )
                              )

def bt_irq(event, data):
    print('bt irq', event, data)

_IRQ_ALL = const(0xffff)
bt.irq(bt_irq, IRQ_ALL)

def adv_encode(adv_type, value):
    return bytes((len(value) + 1, adv_type,)) + value

def adv_encode_name(name):
    return adv_encode(0x09, name.encode())

def adv():
    bt.advertise(100, adv_encode(0x01, b'\x06') + adv_encode(0x03, b'\x0d\x18') + adv_encode(0x19, b'\xc1\x03') + adv_encode_name('esp32hr'))

adv()
```

Then once a phone is connected (get the connection handle from the connection event), you can use

```python
bt.gatts_notify(conn_handle, hr, b'\x00\x40')
```

(where the '\x40` is the heart rate -- see https://www.bluetooth.com/wp-content/uploads/Sitecore-Media-Library/Gatt/Xml/Characteristics/org.bluetooth.characteristic.heart_rate_measurement.xml for the details of this characteristic).

jonsmirl commented:
I am on common-ble-api-esp32-nimble,  I am adding this MDNS wrapper into my local build.
https://github.com/loboris/MicroPython_ESP32_psRAM_LoBo/blob/master/MicroPython_BUILD/components/micropython/esp32/network_mdns.c

I want to see if MDNS and BLE will both run without crashing. I have not had a lot of luck with that.

dpgeorge commented:
> and move the files as suggested

I'm happy to do that during merge if it proves too difficult.

jimmo committed:
stm32: Provide port-specific implementation for Nimble on STM32.

jimmo committed:
extmod/modbluetooth_nimble: Implement modbluetooth API with Nimble.

jimmo commented:
> When merging I would do the following reorganisation of files (let me know if you think this is not a good idea):

I agree. This isn't just a matter of moving the files due to some stm32 specific stuff in nimble/, but this was a good refactoring to do anyway. Done (with some substantial rebasing and rewriting of history, which was worth doing anyway too).

> Should the address be included? Might make sense, and is a breaking API change so better to do it sooner rather than later.

Done. It simplifies the code and reduces code size. I should have just done it this way to start.

> Bytes is covered by mp_bluetooth_parse_uuid_128bit_str so I think this TODO can be removed.

Actually what I meant here is something like `bt.UUID(b'\xff....')` but we can implement that later if necessary.

> That's probably a good idea. What we could do is just make the default trigger=0xffff so it's never required.

Good idea. Done.

jimmo committed:
unix/coverage: Add coverage tests for ringbuf.

jimmo committed:
stm32/mpconfigport.h: Add modbluetooth module to stm32.

dpgeorge commented:
THIS IS FINALLY MERGED!  Thanks to @aykevl and @jimmo for the hard work, and all others who did testing and reviewing.

dmazzella commented:
@dpgeorge any progress on nimble integration on stm32wb?

pacmac commented:
small issue, on the PYBOARD-D, before pairing/connecting the device is advertised as a ESP32 device (dont recall the actual ID), after pairing/connecting appears as 'PYBD'

jimmo commented:
> small issue, on the PYBOARD-D, before pairing/connecting the device is advertised as a ESP32 device (dont recall the actual ID), after pairing/connecting appears as 'PYBD'

If you're using the code snippet above, it will _advertise_ as "esp32hr", but the device name is hardcoded in the firmware to `PYBD`. (and `ESP32` on STM32) which you will see on connection.

I've added an item to #5186 to make this device name configurable.

pacmac commented:
Oops, how careless of me, i copied the code snippet verbatim and did not even notice that

> On 8 Oct 2019, at 07:50, Jim Mussared <notifications@github.com> wrote:
> 
> ﻿
> small issue, on the PYBOARD-D, before pairing/connecting the device is advertised as a ESP32 device (dont recall the actual ID), after pairing/connecting appears as 'PYBD'
> 
> If you're using the code snippet above, it will advertise as "esp32hr", but the device name is hardcoded in the firmware to PYBD. (and ESP32 on STM32) which you will see on connection.
> 
> I've added an item to #5186 to make this device name configurable.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
